### PR TITLE
Implement bitswap in the network behaviour using libp2p_bitswap.

### PR DESCRIPTION
**Pull request by expenses**
I *believe* this would be a good start to implementing #6075. As [libp2p-bitswap](https://github.com/ipfs-rust/libp2p-bitswap) still needs some work (see https://github.com/paritytech/substrate/issues/6075#issuecomment-667079633), this PR probably needs a lot of work in order to be merged. A few points:

~~* Ideally the `Bitswap` behaviour would be wrapped in a `Toggle` so that it's only activated when a chain requests it, but we can't do that until https://github.com/libp2p/rust-libp2p/pull/1684 gets merged.~~
~~* I think we should be inserting the bitswap blocks into the DHT as well, so that nodes we're not directly connected to will be able to access them.~~
~~* As the idea is to allow the fetching of preimages by an IPFS client, I'm not sure if we should be messing about with IPLD as well, or if this is too low level for that.~~

polkadot companion: https://github.com/paritytech/polkadot/pull/1550

### CODE REVIEW

File: client/network/src/service.rs
Comment (tomaka) [inline client/network/src/service.rs:64]: I'd strongly prefer is the network didn't depend on some offchain storage types. Let's please not spaghetti-fy the code even more.

File: client/network/src/service.rs
Comment (tomaka) [inline client/network/src/service.rs:560]: I'm not a fan of exposing private types such as `libp2p_bitswap::Bitswap`. For the sake of any future maintainer, please duplicate the function definitions of bitswap.

File: client/network/src/behaviour.rs
Comment (tomaka) [inline client/network/src/behaviour.rs:54]:
```suggestion
	bitswap: libp2p_bitswap::Bitswap,
```

File: client/network/src/behaviour.rs
Comment (tomaka) [inline client/network/src/behaviour.rs:155]: Please don't expose types from `libp2p_bitswap`

File: client/network/src/behaviour.rs
Comment (tomaka) [inline client/network/src/behaviour.rs:467]:
```suggestion
				self.substrate.add_discovered_nodes(iter::once(peer_id.clone()));
				self.bitswap.connect(peer_id);
```

File: client/network/src/behaviour.rs
Comment (tomaka) [inline client/network/src/behaviour.rs:273]:
```suggestion
```

File: client/network/src/protocol/event.rs
Comment (tomaka) [inline client/network/src/protocol/event.rs:49]: Please don't expose types from `libp2p_bitswap`

**Comment by dvc94ch** [comment]
One problem is that ipfs data will leak, as there isn't a way of disposing of it. And since it's content addressed you'll need to ref count the blocks as the data is deduplicated.

To get "ipld" support you just need to implement the ipld store traits, you can easily do that, and then you get all the ipld codecs and ipld collections for free.

**Comment by dvc94ch** [comment]
I was planning on reimplementing bitswap as a light client notification protocol, but it doesn't seem there is a way to perform the get-providers dht query from the `NetworkService`. This would be an alternative that would also work for us. I'm going to make the entire network implementation of ipfs-embed replaceable so that I can drop in the light client. The db is already shared between ipfs-embed and the light-client.

@tomaka do you think exposing get-provider and start-providing in NetworkService could be a faster way forward?

**Comment by tomaka** [comment]
This PR doesn't do anything related to the DHT yet as far as I can tell, but one thing to be aware of is that we're customizing the name of the Kademlia protocol to include the id of the chain, as a way to prevent nodes from different chains from connecting to each other.

I suppose that it would fundamentally be better to also connected to a shared DHT, or to IPFS itself, rather than to only the DHT of a specific chain. However, this would be problematic as we are already now suffering from problems caused by the high number of TCP sockets.

> a light client notification protocol

Request-response protocols, added in #6634, would be more appropriate I think. They are similar to RPC queries, whereas if you try to emulate request-reponses on top of notifications you will suffer from all sorts of issues.

> @tomaka do you think exposing get-provider and start-providing in NetworkService could be a faster way forward?

That'd be totally reasonable.

**Comment by dvc94ch** [comment]
> This PR doesn't do anything related to the DHT yet as far as I can tell, but one thing to be aware of is that we're customizing the name of the Kademlia protocol to include the id of the chain, as a way to prevent nodes from different chains from connecting to each other.

This is what we want as we're using custom codecs and hashers based on the sp-trie module. We have a prototype "chain" pallet that let's you deal with comonalities between chains of ipfs blocks that store user or team data. The "chain" pallet has a list of "authorities" that are allowed to update the head and prove that the ancestor was the previous head. We found this pattern useful for data that is not interesting for all users, like for example your "twitter feed" and where users can follow one of these "chains" to both get updates and replicate the blocks. In the far future we can maybe leverage a more sophisticated incetivized availability system.

**Comment by expenses** [comment]
Are there any changes that need to be made to this before it can be merged?

**Comment by dvc94ch** [comment]
I guess the dht stuff is missing? Also note that the next version of bitswap will be generic over the multihash code.

For this to be useful to light clients so that they can send and receive blocks from each other, there needs to be some nat traversal support. For simple nat's that should happen automatically once quic is used by default instead of tcp, as it reuses the same port for incoming and outgoing connections. For more annoying nat's there's a PR open to detect if nat traversal is succesful in libp2p https://github.com/libp2p/rust-libp2p/pull/1672, and if it isn't falling back to the relay protocol is necessary, which is also a wip.

File: client/network/src/behaviour.rs
Comment (seunlanlege) [inline client/network/src/behaviour.rs:162]: what's the difference between a `Box<[u8]>` and a `Vec<u8>`

Comment (expenses) [reply client/network/src/behaviour.rs:162]: The main thing is that it's the type that `libp2p_bitswap` uses. Another thing is that you know that all the allocated bytes are used, as opposed to a `Vec` where the allocated `capacity` could be a lot bigger than the `len`.

Comment (dvc94ch) [reply client/network/src/behaviour.rs:162]: I don't think it's true, when converting a Vec to a Box the capacity isn't truncated. The difference to me is that it better represents bytes because most operations on a vec aren't useful when dealing with bytes. For example it may be useful to sort a Vec<u8>, it's probably not useful in most cases to sort a `Box<[u8]>`. It also let's you implement a trait for a Vec<T> and a Box<[u8]> as sometimes you wan't to handle bytes differently from a list. But I've mostly given up on this, as everyone just uses Vec<u8>.

Comment (seunlanlege) [reply client/network/src/behaviour.rs:162]: > as opposed to a Vec where the allocated capacity could be a lot bigger than the len.

yeah that makes sense.

Comment (expenses) [reply client/network/src/behaviour.rs:162]: @dvc94ch [`into_boxed_slice`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.into_boxed_slice) drops the excess capacity, at least.

Comment (seunlanlege) [reply client/network/src/behaviour.rs:162]: if you want something that better represents bytes, why not just use the `bytes` crate?

Comment (dvc94ch) [reply client/network/src/behaviour.rs:162]: What's the advantage of using the bytes crate vs a Box<[u8]>?

Comment (seunlanlege) [reply client/network/src/behaviour.rs:162]: > What's the advantage of using the bytes crate vs a Box<[u8]>?

reference-counted so cheaper to clone.

Comment (seunlanlege) [reply client/network/src/behaviour.rs:162]: I don't have any strong opinions on this topic, I just have never seen a `Box<[u8]>` in the wild before and I thought it was a very clever way to abstract bytes.

Comment (dvc94ch) [reply client/network/src/behaviour.rs:162]: So what you want is an Arc<[u8]>. Anyway looks like into_boxed_slice does drop excess capacity, there used to be some issues with it https://github.com/rust-lang/rust/issues/36284, but I can't find why I thought it didn't.

Comment (seunlanlege) [reply client/network/src/behaviour.rs:162]: also, I just checked and the bytes crate has a lot of convenience methods for working with bytes :smile:  (LE/BE get/set).

Comment (dvc94ch) [reply client/network/src/behaviour.rs:162]: As I said, I might change it to just return a Vec<u8>, if people prefer. I've had a few people disagree on the utility of a Box<[u8]>. I probably won't force people to use a specific crate, you can easily convert a vec or a box to bytes.

Comment (tomaka) [reply client/network/src/behaviour.rs:162]: In this kind of situation, in my opinion, one should simply expose implementation details. If the implementation uses `Vec`, return a `Vec`, so that you're not copying bytes for no reason.

**Comment by expenses** [comment]
Waiting on https://github.com/ipfs-rust/rust-ipld/pull/47.

**Comment by expenses** [comment]
Waiting on https://github.com/async-rs/async-std/pull/870.

**Review by dvc94ch — COMMENTED** [review]
So I reviewed the code again. There are a couple of issues with the current implementation. The changes aren't atomic so if there is a crash half way through inserting a batch for example, the store will be in an inconsistent state. Also I noticed that the light client keeps the blocks in memory and they aren't persisted to disk.

Maybe substrate shouldn't implement the store, and instead just expose dht and bitswap operations via the Network thing.

**Comment by expenses** [comment]
@dvc94ch This is ready for a re-review (the button has stopped working for me :woman_shrugging: )

File: client/service/src/builder.rs
Comment (dvc94ch) [inline client/service/src/builder.rs:771]: typo

File: client/service/src/builder.rs
Comment (dvc94ch) [inline client/service/src/builder.rs:778]: Can this not be fixed to `DefaultStoreParams`?

File: client/service/src/lib.rs
Comment (dvc94ch) [inline client/service/src/lib.rs:381]: ipfs-embed moved away from get making any changes to the stored blocks. in ipfs-embed this will actually not insert anything at all as the block isn't pinned. I see the need for recording the timestamps and having a mixture of pins, refcounts and timestamps determine if a block is live or not. That way the store can be used as a cache without leading to storage leakage.

File: client/service/src/lib.rs
Comment (dvc94ch) [inline client/service/src/lib.rs:386]: get will potentially try to fetch the block from the network. I think you only want to respond if you have the block locally. If the block isn't available locally or found on the network it will return a `BlockNotFound`. In general there is a problem that the semantics of go-ipfs, js-ipfs are not clear (or outright insufficient), and while formulating and implementing precise semantics for ipfs-embed in terms of atomicity/durability/garbage collection, they are still in flux as more experience is gained building applications on top. That's why I'd recommend completely removing any storage logic for now.

Comment (expenses) [reply client/service/src/lib.rs:386]: So should I just remove `handle_bitswap_event` and `IpldStoreBuilder` entirely?

Comment (dvc94ch) [reply client/service/src/lib.rs:386]: Not sure how you intend to use it. While exchanging blocks with full nodes and validators could be interesting if light clients can pay them to store blocks for them, for now I'm mainly interested in using it for light client to light client data exchange. So I think removing those entirely would be fine for now.

**Review by dvc94ch — APPROVED** [review]
LGTM! Thanks.

**Comment by gnunicorn** [comment]
as this is mainly in the network, I think @tomaka should have something to say about this...

File: client/network/src/service.rs
Comment (tomaka) [inline client/network/src/service.rs:1072]: This needs to be documented.
I don't know how bitswap works, and that's great because it should in my opinion be possible for someone who doesn't know how bitswap works to understand this API.

What's the general workflow of the thing? What happens if `send_block` is called on a peer that doesn't want this block?

File: client/network/src/behaviour.rs
Comment (tomaka) [inline client/network/src/behaviour.rs:186]: Can this be called multiple times with the same block? What's the relationship between the `Cid` and the data? Is it guaranteed to be in response to a WANT request?

**Review by tomaka — CHANGES_REQUESTED** [review]

File: client/network/src/gossip.rs
Comment (expenses) [reply client/network/src/gossip.rs]: Line 1928 of https://gitlab.parity.io/parity/substrate/-/jobs/658201#L1928 is why we need this.

File: client/network/src/behaviour.rs
Comment (tomaka) [inline client/network/src/behaviour.rs:189]: That sounds like an attack vector.
@dvc94ch Can a node just connect and send, say, 2 GiB of data without us having asked for it, and we will simply buffer it?

Comment (expenses) [reply client/network/src/behaviour.rs:189]: This should probably be combined with the reputation system right?

Comment (dvc94ch) [reply client/network/src/behaviour.rs:189]: no this can't happen https://github.com/ipfs-rust/tiny-multihash/issues/12

File: client/network/src/behaviour.rs
Comment (dvc94ch) [reply client/network/src/behaviour.rs]: yes

File: client/network/src/service.rs
Comment (dvc94ch) [reply client/network/src/service.rs]: bitswap_send_block is usually in response to a want event. the block is added to a queue and if a cancel request is received before the message is sent it will be removed from the queue. reading the code, if the cancel request comes in before the want request was processed, the cancel request will be ignored. I guess it could use some work, but it's not "terrible"

File: client/network/src/service.rs
Comment (dvc94ch) [reply client/network/src/service.rs]: So these are open connections to other peers. It will send want requests to any peer that connects that speaks bitswap. For each open connection it maintains a list of blocks the other peer wants. If in the future the block is available it sends it to the peer. This architecture is because when I first wrote bitswap it seemed like you could have a strategy that gets a block from a friendly peer to get a block from a new peer. Now I think it should probably just ignore the want requests of blocks it doesn't have.

File: client/network/src/behaviour.rs
Comment (expenses) [reply client/network/src/behaviour.rs:189]: @dvc94ch right, but that 2 GiB of data is still held in memory for a period of time, right?

Comment (dvc94ch) [reply client/network/src/behaviour.rs:189]: Well that depends on the transport. A message is processed from the queue, and dropped if it wasn't requested. We don't have a sophisticated mechanism for disconnecting from peers, I thought that substrate handles bandwidth limiting, telemetry and reputation, and instead of handling this stuff myself in bitswap/ipfs-embed, I thought we can leverage that. I'm glad to redesign bitswap if there is se input on how it should work

File: client/network/src/behaviour.rs
Comment (dvc94ch) [reply client/network/src/behaviour.rs]: > Can this be called multiple times with the same block?

no, unless the events weren't polled until the block arrived and a new want request was submitted.
> What's the relationship between the Cid and the data

the cid hash matches the data, the cid is the one submitted in the want request
>  Is it guaranteed to be in response to a WANT request?

it's guaranteed to be in response to a want request.

**Comment by dvc94ch** [comment]
So where do we go from here? I guess I could pr libp2p-bitswap to rust-libp2p, then we could review that first before moving on to the substrate specifics?

File: client/network/src/behaviour.rs
Comment (dvc94ch) [reply client/network/src/behaviour.rs:189]: FYI didn't even see the other comment on cid and data don't match. The bitswap never sends a hash, only the data. The cid is reconstructed while parsing the message, they are guaranteed to match https://github.com/ipfs-rust/libp2p-bitswap/blob/23f67ea8e36230aac4ef5c1feef51af6f157df38/src/message.rs#L155

**Comment by dvc94ch** [comment]
@expenses it seems the docs are still incorrect. If you're busy with other stuff, feel free to close the pr and I'll merge the changes that make sense into mine.

**Comment by expenses** [comment]
@dvc94ch ok, that makes sense :+1:

File: client/network/src/behaviour.rs
Comment (tomaka) [reply client/network/src/behaviour.rs:189]: >  A message is processed from the queue, and dropped if it wasn't requested. We don't have a sophisticated mechanism for disconnecting from peers, I thought that substrate handles bandwidth limiting, telemetry and reputation, and instead of handling this stuff myself in bitswap/ipfs-embed, I thought we can leverage that. I'm glad to redesign bitswap if there is some input on how it should work

I'm not familiar with the bitswap protocol or the internals of bitswap, so I'm just naively asking.
The problem I have in mind is: someone can just connect, open 1000 substreams (or whatever the maximum is), negotiate bitswap, send a length prefix of `N`, and then send `N - 1` bytes but never the last byte. Do this multiple times simultaneously and you can probably make the node run out of memory.
This is entirely specific to the bitswap protocol, and not something Substrate can handle.

What we do for other protocols is enforce a limit of one substream per protocol per connection, and make sure that the remote can never send a large amount of data without having explicitly requested it first.

**Comment by tomaka** [comment]
Link for note keeping: https://github.com/paritytech/substrate/pull/7133

File: client/network/src/behaviour.rs
Comment (dvc94ch) [reply client/network/src/behaviour.rs:189]: > What we do for other protocols is enforce a limit of one substream per protocol per connection

https://github.com/ipfs-rust/libp2p-bitswap/issues/7

> make sure that the remote can never send a large amount of data without having explicitly requested it first.

I think this is already handled. Does this satisfy the requirement? https://github.com/ipfs-rust/libp2p-bitswap/blob/master/src/behaviour.rs#L216 Or is more work needed?

Comment (tomaka) [reply client/network/src/behaviour.rs:189]: > can you point me to a specific protocol that handles this, so I can look how it's done?

I'm not familiar with this code unfortunately, but [libp2p's request-response should do that](https://github.com/libp2p/rust-libp2p/tree/master/protocols/request-response/src)

> I think this is already handled. Does this satisfy the requirement? https://github.com/ipfs-rust/libp2p-bitswap/blob/master/src/behaviour.rs#L216 Or is more work needed?

As far as I understand, you're still entirely buffering the message, and then only discard it, so it really depends on the maximum size of the message. If it's 512 kiB it's ok, if it's 2 GiB it's not.

Comment (dvc94ch) [reply client/network/src/behaviour.rs:189]: I think these issues are handled much better now https://github.com/ipfs-rust/libp2p-bitswap/pull/8.

@tomaka how concerned are you with interopability with other implementations? In particular now we avoid packing multiple requests responses in the same message, and reuse the substream to send multiple messages. Also even if we add that supporting bitswap 1.1.0 and 1.0.0 is not something I'm particularly interested in working on, and the api is really focused around bitswap 1.2.0 which works completely different from previous versions.

Comment (tomaka) [reply client/network/src/behaviour.rs:189]: > how concerned are you with interopability with other implementations?

If this feature is used specifically for large files storage using Substrate, I guess we don't really care.

**Pull request closed without merging**
